test/unit/compare-tree/compare-tree.test.js

```diff
@@ -0,0 +1,330 @@
+    })
+
+    it('does not remove parents with other compares', function () {
+      let handler = jest.fn()
+      let query = this.tree.on('meta.selected', handler)
+
+      this.tree.on('meta.focused', handler)
+      this.tree.off('meta.selected', handler)
+
+      expect(this.tree.nodes[query.id]).toBeUndefined()
+      expect(this.tree.nodes['meta.selected']).toBeUndefined()
+      expect(this.tree.nodes['meta.focused']).toBeDefined()
+      expect(this.tree.nodes['meta']).toBeDefined()
+    })
+
+    it('keeps the root node in tact', function () {
+      let handler = jest.fn()
+
+      this.tree.on('meta.selected', handler)
+
+      // 'root', 'meta', 'selected', 'handler'
+      expect(Object.keys(this.tree.nodes)).toHaveLength(4)
+
+      this.tree.off('meta.selected', handler)
+
+      expect(Object.keys(this.tree.nodes)).toEqual(['~'])
+    })
+
+    it('keeps a query if it still has compares left', function () {
+      let one = jest.fn()
+      let two = jest.fn()
+
+      let next = set(this.tree.state, ['meta', 'selected'], {})
+      let query = this.tree.on('meta.selected', one)
+
+      this.tree.on('meta.selected', two)
+
+      this.tree.off('meta.selected', one)
+
+      this.tree.update(next)
+
+      expect(one).not.toHaveBeenCalled()
+      expect(two).toHaveBeenCalled()
+
+      expect(this.tree.nodes[query.id]).toBeDefined()
+    })
+
+    it('gracefully handles parents missing an edge (which should never happen)', function () {
+      let one = n => n
+      let two = n => n
+
+      this.tree.on('a.b.c', one)
+
+      let query = this.tree.on('a.b', two)
+
+      let aB = this.tree.nodes['a.b.c']
+      aB.edges = aB.edges.filter(e => e === query)
+
+      this.tree.off('a.b.c', one)
+
+      let keys = Object.keys(this.tree.nodes)
+
+      expect(keys).toContain('a')
+      expect(keys).toContain('a.b')
+      expect(keys).toContain('query:a.b')
+      expect(keys).not.toContain('query:a.b.c')
+    })
+
+    it('gracefully handles compares that do not exist', function () {
+      this.tree.on('a.b.c', n => n)
+      this.tree.off('a.b.c', n => n)
+
+      let keys = Object.keys(this.tree.nodes)
+
+      expect(keys).toContain('a')
+      expect(keys).toContain('a.b')
+      expect(keys).toContain('a.b.c')
+      expect(keys).toContain('query:a.b.c')
+    })
+  })
+
+  describe('::connect', function () {
+
+    it('edges can not connect to themselves', function () {
+      let node = this.tree.add('node')
+
+      node.connect(node)
+
+      expect(node.edges).not.toContain(node)
+    })
+
+  })
+
+  describe('::update', function () {
+    it('can traverse compares for missing keys', function () {
+      this.tree.on('a.b.c', n => n)
+      this.tree.update({})
+    })
+
+    it('can traverse missing state keys', function () {
+      this.tree.on('meta.selected', n => n)
+
+      this.tree.update({ meta: null })
+
+      this.tree.update({
+        meta: { selected: true }
+      })
```

Comment:
ah, sorry, the assertion here is that nothing blows up, since the `meta.selected` path disappears, and then reappears